### PR TITLE
Add flow mirror point delete FSM

### PR DESCRIPTION
**Pull request by dpoltavets**
Related to: #3849

**MERGE NOTE: DO NOT MERGE BEFORE MERGING #4232**

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/DeallocateFlowMirrorPathResourcesAction.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        FlowMirrorPath flowMirrorPath = flowMirrorPathRepository.findById(mirrorPathId)
+                .orElseThrow(() -> new FlowProcessingException(ErrorType.NOT_FOUND,
+                        format("Flow mirror point %s not found", mirrorPathId)));
+
+        FlowMirrorPoints flowMirrorPoints = flowMirrorPath.getFlowMirrorPoints();
+        stateMachine.setFlowPathId(flowMirrorPoints.getFlowPathId());
+        stateMachine.setMirrorSwitchId(flowMirrorPoints.getMirrorSwitchId());
+
+        resourcesManager.deallocateCookie(flowMirrorPath.getCookie().getFlowEffectiveId());
```
Comment (surabujin) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/DeallocateFlowMirrorPathResourcesAction.java:62]: Do we share cookie with target flow? If so, how we can release it on mirror point removal (while target flow still exists)?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/ValidateRequestAction.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            Flow foundFlow = getFlow(flowId);
+            if (foundFlow.getStatus() == FlowStatus.IN_PROGRESS) {
+                throw new FlowProcessingException(ErrorType.REQUEST_INVALID,
+                        format("Flow %s is in progress now", flowId));
+            }
+
+            FlowMirrorPath flowMirrorPath = flowMirrorPathRepository.findById(mirrorPathId)
+                    .orElseThrow(() -> new FlowProcessingException(ErrorType.NOT_FOUND,
+                            format("Flow mirror point %s not found", mirrorPathId)));
+            if (flowMirrorPath.getStatus() == FlowPathStatus.IN_PROGRESS) {
```
Comment (surabujin) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/ValidateRequestAction.java:80]: What will happen if target flow will be changed while we are performing some action on mirror point?

**Comment by surabujin** [comment]
There is no FSM diagramm.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/DeallocateFlowMirrorPathResourcesAction.java (hunk 1 above)
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/DeallocateFlowMirrorPathResourcesAction.java:62]: We do not share cookie with target flow.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/ValidateRequestAction.java (hunk 2 above)
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/ValidateRequestAction.java:83]: Now the flow is set to IN_PROGRESS status

**Comment by dpoltavets** [comment]
FSM diagram has been added

**Review by niksv — APPROVED** [review]

**Review by surabujin — CHANGES_REQUESTED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/EmitCommandRequestsAction.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        FlowCommandBuilder commandBuilder = commandBuilderFactory.getBuilder(flow.getEncapsulationType());
+        Collection<FlowSegmentRequestFactory> commands =
+                buildCommands(commandBuilder, stateMachine, flow, mirrorPoints);
+
+        // emitting
+        SpeakerRequestEmitter requestEmitter;
+        if (mirrorPoints.getMirrorPaths().isEmpty()) {
+            requestEmitter = SpeakerRemoveSegmentEmitter.INSTANCE;
+        } else {
+            requestEmitter = SpeakerInstallSegmentEmitter.INSTANCE;
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/EmitCommandRequestsAction.java:86]: As agreed offline, sending commands for installation is necessary to modify an existing group, if not all mirror points have been removed.

**Review by surabujin — APPROVED** [review]

**Pull request merged**